src/main/java/com/artipie/asto/s3/S3Storage.java

```diff
@@ -181,6 +181,31 @@ public S3Storage(final S3AsyncClient client, final String bucket) {
         return promise;
     }
 
+    @Override
+    public CompletableFuture<Void> delete(final Key key) {
+        return this.exists(key).thenCompose(
+            exists -> {
+                final CompletableFuture<Void> deleted;
+                if (exists) {
+                    deleted = this.client.deleteObject(
+                        DeleteObjectRequest.builder()
+                            .bucket(this.bucket)
+                            .key(key.string())
+                            .build()
+                    ).thenCompose(
+                        response -> CompletableFuture.allOf()
+                    );
+                } else {
```

Comment:
@olegmoz do we need to handle this case at asto side? Maybe S3 client throws similar exception if S3 object can not be found?

Reply:
@g4s8 sadly, the client does not report absence of key. I checked it with S3 emulator and read through official API docs